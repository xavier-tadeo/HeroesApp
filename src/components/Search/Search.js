import { useLocation, useNavigate } from "react-router-dom";
import queryString from "query-string";
import { getHerosByName } from "../../helpers/getHerosByName";
import { useForm } from "../../hooks/useForm";
import { HeroCard } from "../HeroCard/HeroCard";
import { useMemo } from "react";

export const Search = () => {
  const navigate = useNavigate();
  const location = useLocation();

  const { q = "" } = queryString.parse(location.search);

  const [values, handleInputChange] = useForm({
    searchText: q,
  });

  const { searchText } = values;

  const heroFilter = useMemo(() => getHerosByName(q), [q]);

  const handleSubmit = (evt) => {
    evt.preventDefault();
    navigate(`?q=${searchText}`);
  };

  return (
    <>
      <h1>Search</h1>
      <hr />
      <div className="row">
        <div className="col-5">
          <hr />
          <form onSubmit={handleSubmit}>
            <input
              type="text"
              placeholder="Search hero"
              className="form-control"
              name="searchText"
              autoComplete="off"
              onChange={handleInputChange}
              value={searchText}
            />
            <button className="btn btn-success mt-2" type="submit">
              Search...
            </button>
          </form>
        </div>
        <div className="col-7">
          <h4>Results</h4>
          <hr />

          {!q ? (
            <div className="alert alert-info">Search Hero</div>
          ) : (
            heroFilter?.length === 0 && (
              <div className="alert alert-danger">Don't have results: {q}</div>
            )
          )}

          {heroFilter?.map((hero) => (
            <HeroCard key={hero.id} {...hero} />
          ))}
        </div>
      </div>
    </>
  );
};
