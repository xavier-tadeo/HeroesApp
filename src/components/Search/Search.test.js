import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import ReactTestRenderer from "react-test-renderer";
import { MemoryRouter } from "react-router-dom";
import { Search } from "./Search";

const mockNavigate = jest.fn();

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useNavigate: () => mockNavigate,
}));

describe("Give a Search component", () => {
  test("Should rendered correcty", () => {
    const view = ReactTestRenderer.create(
      <MemoryRouter initialEntries={["/search"]}>
        <Search />
      </MemoryRouter>
    );

    expect(view.toJSON()).toMatchSnapshot();
  });

  test("Should have a text 'Search Hero' in a document", () => {
    render(
      <MemoryRouter initialEntries={["/search"]}>
        <Search />
      </MemoryRouter>
    );

    const searchText = screen.getByText("Search Hero");

    expect(searchText).toBeInTheDocument();
  });

  test("Should have show a Batman and input with this value", () => {
    render(
      <MemoryRouter initialEntries={["/search?q=batman"]}>
        <Search />
      </MemoryRouter>
    );

    const input = screen.getByPlaceholderText("Search hero");

    expect(input.value).toBe("batman");
  });

  test("It should show error if not found hero", () => {
    render(
      <MemoryRouter initialEntries={["/search?q=batman123"]}>
        <Search />
      </MemoryRouter>
    );

    const expectText = screen.getByText("Don't have results: batman123");

    expect(expectText).toBeInTheDocument();
  });

  describe("When click a button with info in input", () => {
    test("Then it should call the function handleSubmit", () => {
      render(
        <MemoryRouter initialEntries={["/search"]}>
          <Search />
        </MemoryRouter>
      );

      const input = screen.getByPlaceholderText("Search hero");
      const button = screen.getByRole("button");

      userEvent.type(input, "Batman");
      userEvent.click(button);

      expect(mockNavigate).toHaveBeenCalledWith("?q=Batman");
    });
  });
});
