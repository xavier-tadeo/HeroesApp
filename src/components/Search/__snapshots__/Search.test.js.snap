// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Give a Search component Should rendered correcty 1`] = `
Array [
  <h1>
    Search
  </h1>,
  <hr />,
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Search hero"
          type="text"
          value=""
        />
        <button
          className="btn btn-success mt-2"
          type="submit"
        >
          Search...
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="alert alert-info"
      >
        Search Hero
      </div>
    </div>
  </div>,
]
`;
